pkg/daemon/controller/controller.go

```diff
@@ -265,6 +276,79 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
+			return false, fmt.Errorf("error setting reboot annotation on node %s: %w", sc.nodeName, err)
+		}
+		return true, nil
+	}
+	// Wait until the environment variables at the process level are set as expected.
+	// This will only be picked up after WMCO reboots the instance
+	err = wait.PollUntilContextTimeout(sc.ctx, 15*time.Second, 5*time.Minute, true,
+		func(ctx context.Context) (done bool, err error) {
+			stillNeedsReboot := false
+			for _, varName := range cluster.SupportedProxyVars {
+				cmd := fmt.Sprintf("[Environment]::GetEnvironmentVariable('%s', 'Process')", varName)
+				out, err := sc.psCmdRunner.Run(cmd)
+				if err != nil {
+					return false, fmt.Errorf("error running PowerShell command %s with output %s: %w", cmd, out, err)
+				}
+				if strings.TrimSpace(out) != envVars[varName] {
+					stillNeedsReboot = true
+				}
+			}
+			return !stillNeedsReboot, nil
+		})
+	if err != nil {
+		return true, fmt.Errorf("error waiting for environment vars to get picked up by processes: %w", err)
+	}
+	// Remove the reboot annotation after we know the reboot occurred successfully. No-op if already not present
+	return false, metadata.RemoveRebootAnnotation(sc.ctx, sc.client, node)
+}
+
+// ensureVarsAreUpToDate ensures that the proxy environment variables are set as expected on the instance
+// If there's any changes, an instance restart is required to ensure all processes pick up the updated values.
+func (sc *ServiceController) ensureVarsAreUpToDate(envVars map[string]string) (bool, error) {
+	// systemEnvVarRegistryPath is where system level environment variables are stored in the Windows OS
+	const systemEnvVarRegistryPath = `SYSTEM\CurrentControlSet\Control\Session Manager\Environment`
+	restartRequired := false
+	for key, expectedVal := range envVars {
+		registryKey, err := registry.OpenKey(registry.LOCAL_MACHINE, systemEnvVarRegistryPath, registry.ALL_ACCESS)
+		if err != nil {
+			return false, fmt.Errorf("unable to open Windows system registry key %s: %w", systemEnvVarRegistryPath, err)
+		}
+		actualVal, _, err := registryKey.GetStringValue(key)
+		if err != nil && err != registry.ErrNotExist {
+			return false, fmt.Errorf("unable to read environment variable %s: %w", key, err)
+		}
+
+		if actualVal != expectedVal {
+			klog.Infof("updating environment variable %s", key)
+			// Because we modify env vars are the "system" level rather than the ephemeral "process" level,
+			// we cannot use os.Setenv, which is a wrapper for syscall.SetEnvironmentVariable
+			// As per Microsoft docs: "Calling SetEnvironmentVariable has no effect on the system environment variables"
+			err := registryKey.SetStringValue(key, expectedVal)
+			if err != nil {
+				// Do not log value as proxy information is sensitive
+				return false, fmt.Errorf("unable to set environment variable %s: %w", key, err)
+			}
+			restartRequired = true
+		}
+		if err := registryKey.Close(); err != nil {
```

Comment:
Don't you need to always close this key in case of errors above?

Reply:
Yes, it should always be closed. We had changed it based on [this comment](https://github.com/openshift/windows-machine-config-operator/pull/1536#discussion_r1223540487), but now I am leaning towards doing back to `defer registryKey.Close()`

Reply:
You can go back to `defer registryKey.Close()` but you need to catch the error and not just ignore it. I thought the PR was updated to do that correctly but moved back.

Reply:
Sorry I don't understand the ask. How should we catch/return the error in a defer call? Whatever is returned from a defer func will not affect the function it is called from, ex: https://go.dev/play/p/bXAnQkdk517

Reply:
The [sys/windows/registry package](https://pkg.go.dev/golang.org/x/sys/windows/registry) itself recommends just deferring the close
```
key, err := registry.OpenKey(...)
if err != nil {
    return err
}
defer key.Close()
```